File: Viex.WebSockets.Api/Hubs/WaitingRoomHub.cs
using Microsoft.AspNetCore.SignalR;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Timers;
using Viex.WebSockets.Api.Services;
using Viex.WebSockets.Core.Payloads;

namespace Viex.WebSockets.Api.Hubs
{
    public partial class ViexHub
    {
        public async Task JoinWaitingRoom(JoinWaitingRoomPayload payload)
        {
            await Groups.AddToGroupAsync(Context.ConnectionId, payload.RoomPassword);
            await Clients.Group(payload.RoomPassword).SendAsync("GuestJoinedWaitingRoom", payload.Username);
        }

        public async Task LeaveWaitingRoom(LeaveWaitingRoomPayload payload)
        {
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, payload.RoomPassword);
            await Clients.Group(payload.RoomPassword).SendAsync("GuestLeftWaitingRoom", payload.Username);
        }

        public Task StartCountDown(string waitingRoomPassword)
        {
            _waitingRoomCountDownTasks.EnqueueJob(waitingRoomPassword);
            return Task.CompletedTask;
        }
    }
}

File: Viex.WebSockets.Web/src/hubs/WaitingRoomHub.ts
import { JoinWaitingRoomPayload } from "@/models/payloads/JoinWaitingRoomPayload";
import { LeaveWaitingRoomPayload } from "@/models/payloads/LeaveWaitingRoomPayload";
import { HubListener } from ".";
import connection from "./connection";

export default {
    join(payload: JoinWaitingRoomPayload) {
        payload.roomPassword = `${payload.roomPassword}`;
        return connection.invoke("JoinWaitingRoom", payload);
    },

    leave(payload: LeaveWaitingRoomPayload) {
        payload.roomPassword = `${payload.roomPassword}`;
        return connection.invoke("LeaveWaitingRoom", payload);
    },

    startCountDown(waitingRoomPassword: string) {
        return connection.invoke("StartCountDown", waitingRoomPassword);
    },

    onGuestJoinedWaitingRoom(listener: HubListener<string>) {
        connection.on("GuestJoinedWaitingRoom", listener);
    },

    onGuestLeftWaitingRoom(listener: HubListener<string>) {
        connection.on("GuestLeftWaitingRoom", listener);
    },

    onRemainingSecondsElapsed(listener: HubListener<number>) {
        connection.on("WaitingRoomRemainingSecondsElapsed", listener);
    },

    onTimeUp(listener: HubListener<void>) {
        connection.on("WaitingRoomTimeUp", listener);
    },
};
